src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -704,6 +704,63 @@ private ImmutableArray<ModifierInfo<TypeSymbol>> DecodeModifiersOrThrow(ref Blob
             return modifiers?.ToImmutableAndFree() ?? default(ImmutableArray<ModifierInfo<TypeSymbol>>);
         }
 
+        /// <exception cref="UnsupportedSignatureContent">If the encoded type is invalid.</exception>
+        /// <exception cref="BadImageFormatException">An exception from metadata reader.</exception>
+        private ImmutableArray<ModifierInfo<TypeSymbol>> DecodeModifiersOrThrow(
+            ref BlobReader signatureReader,
+            out SignatureTypeCode typeCode,
+            Func<TypeSymbol, bool> acceptRequiredModifier,
+            out bool requiredModifierFound)
+        {
+            requiredModifierFound = false;
+            ArrayBuilder<ModifierInfo<TypeSymbol>> modifiers = null;
+
+            for (;;)
+            {
+                typeCode = signatureReader.ReadSignatureTypeCode();
+                bool isOptional;
+
+                if (typeCode == SignatureTypeCode.RequiredModifier)
+                {
+                    isOptional = false;
+                }
+                else if (typeCode == SignatureTypeCode.OptionalModifier)
+                {
+                    isOptional = true;
+                }
+                else
+                {
+                    break;
+                }
+
+                TypeSymbol type = DecodeModifierTypeOrThrow(ref signatureReader);
+
+                // if it is a required modifier, make sure the compiler supports this modifier type
+                if (!isOptional)
+                {
+                    if (acceptRequiredModifier(type))
+                    {
+                        requiredModifierFound = true;
+                    }
+                    else
+                    {
+                        throw new UnsupportedSignatureContent();
+                    }
+                }
+
+                ModifierInfo<TypeSymbol> modifier = new ModifierInfo<TypeSymbol>(isOptional, type);
+
+                if (modifiers == null)
+                {
+                    modifiers = ArrayBuilder<ModifierInfo<TypeSymbol>>.GetInstance();
+                }
+
+                modifiers.Add(modifier);
+            }
+
+            return modifiers?.ToImmutableAndFree() ?? default(ImmutableArray<ModifierInfo<TypeSymbol>>);
+        }
```

Comment:
Consider merging this method with the existing `DecodeModifiersOrThrow`, perhaps by having this method take optional parameters or calling this method from the existing method.